server/test/controllers/api/ProgramApplicationsApiControllerTest.java

```diff
@@ -0,0 +1,142 @@
+package controllers.api;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertThrows;
+import static play.api.test.Helpers.testServerPort;
+import static play.test.Helpers.contentAsString;
+import static play.test.Helpers.fakeRequest;
+import static play.test.Helpers.route;
+
+import auth.ApiKeyGrants;
+import auth.UnauthorizedApiRequestException;
+import com.jayway.jsonpath.DocumentContext;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Optional;
+import models.ApiKey;
+import models.Program;
+import org.apache.http.HttpStatus;
+import org.junit.Before;
+import org.junit.Test;
+import play.mvc.Http;
+import play.mvc.Result;
+import services.applicant.JsonPathProvider;
+import services.export.AbstractExporterTest;
+
+public class ProgramApplicationsApiControllerTest extends AbstractExporterTest {
+
+  private ApiKey apiKey;
+  private static final String keyId = "key-id";
+  private static final String keySecret = "key-secret";
+  private static final String rawCredentials = keyId + ":" + keySecret;
+  private static final String serializedApiKey =
+      Base64.getEncoder().encodeToString(rawCredentials.getBytes(StandardCharsets.UTF_8));
+
+  @Before
+  public void setUp() {
+    apiKey = resourceCreator.createActiveApiKey("test-key", keyId, keySecret);
+  }
+
+  @Test
+  public void list_success_dateRange() {
+    createFakeProgramWithEnumerator();
+    apiKey
+        .getGrants()
+        .grantProgramPermission(fakeProgramWithEnumerator.getSlug(), ApiKeyGrants.Permission.READ);
+    apiKey.save();
+
+    String requestUrl =
+        controllers.api.routes.ProgramApplicationsApiController.list(
+                fakeProgramWithEnumerator.getSlug(),
+                /* fromDate= */ Optional.of("2022-01-15"),
+                /* toDate= */ Optional.of("2022-02-15"),
+                /* nextPageToken= */ Optional.empty(),
+                /* pageSize= */ Optional.empty())
+            .url();
+
+    Result result = doRequest(requestUrl);
+    assertThat(result.status()).isEqualTo(HttpStatus.SC_OK);
+
+    DocumentContext resultJson = JsonPathProvider.getJsonPath().parse(contentAsString(result));
+    assertThat(resultJson.read("payload[0].application_id", Long.class))
+        .isEqualTo(applicationTwo.id);
```

Comment:
Is the connection with the program returned and applicationTwo magic?

I don't see it mentioned in the test otherwise, or understand why it's applicationTwo and not applicationOne.  Can we make it less magic / errorprone ? :D

Reply:
Added a docstring for `createFakeProgramWithEnumerator`

Reply:
The test setup is still quite unreadable.

After looking at the createFake.. method I see that the magic is that each application has a timestamp of 2022-01-01, 2022-02-01, 2022-03-01.

That is hidden by the parent setup and not communicated by the object names, such that searching for [2022-01-15, 2022-02-15) has no readable result from this test.

A few ways to shore this up.
1. Comment on the relative order, "the three applicationX are a month apart successively", and use the actual timestamp of applicationTwo in specifying the fromDate/toDate. EG applicationTwo.submittion_time - 1 ---> toString format

2. Create alias/reference variables in this class from the parents variables, with clearer names; eg
applicationOldest, applicationMiddle, applicationYoungest Those then will convey the relevant distinguishing characteristic.
Having the search time directly associated with the resultant application would still be good here, instead of the magic values.

3. Just document the hidden timestamp setup as it's directly pertinent here. If we didn't have that setup, and we were creating the programs through utility method, the salient thing that would be included in this test are those dates, so we should communicate that information one way or another.

Reply:
Tried to do a combination of better variable names with an explanatory comment in the setup section. I balked at doing date math because it gets weird doing math on instants and then converting them to dates IMO.